### DISCUSSION TITLE
Proposal Async "while" syntactic sugar

### DISCUSSION THREAD

**Discussion by BrianCArnold** [dotnet/csharplang#2599] [Language Ideas]
Currently, to fire an async task and perform actions while it is completing, one usually does this:

var valueRetrievalTask = RetrieveValueAsync(someRelatedValue));
SynchronousMethod1();
SynchronousMethod2();
var retrievedValue = await valueRetrievalTask;

This requires that a task variable, and separates the assignment from the invocation of the async task.

I propose syntax similar to the following: 

var retrievedValue = while(RetrieveValueAsync(someRelatedValue))
{
  SynchronousMethod1();
  SynchronousMethod2();
};

This allows the developer to see the assignment directly next to the invocation, and decoratively associates what executes during the async task.

**Comment by CyrusNajmabadi** [comment]
`while` would be somewhat undesirable as it conveys 'looping'.  Something which is not happening here.

> This allows the developer to see the assignment directly next to the invocation,

Note: you get this even with the original code.  i.e.:

`var valueRetrievalTask = RetrieveValueAsync(someRelatedValue));`

The variable, assignment, and invocation are all grouped together.

**Comment by svick** [comment]
I think you could get fairly close to your desired syntax by using a helper method and a lambda:

```c#
var retrievedValue = await Meanwhile(RetrieveValueAsync(someRelatedValue), () =>
{
    SynchronousMethod1();
    SynchronousMethod2();
});
```

The implementation of `Meanwhile` would be:

```c#
static Task<T> Meanwhile<T>(Task<T> task, Action meanwhileAction)
{
    meanwhileAction();
    return task;
}
```
